modules/auxiliary/scanner/smb/smb_version.rb

```diff
@@ -39,167 +50,250 @@ def smb_direct
+          end
+          info[:capabilities][:encryption] = simple.client.server_encryption_algorithms.map do |algorithm|
+            RubySMB::SMB2::EncryptionCapabilities::ENCRYPTION_ALGORITHM_MAP[algorithm]
+          end
+        end
+        # assume that if the server supports multiple versions, the preferred
+        # dialect will correspond to the latest version
+        preferred_dialect = SMB2_DIALECT_STRINGS[preferred_dialect]
 
-        #
-        # Create the note hash for fingerprint.match
-        #
-        match_conf = { }
+        if simple.client.server_start_time && simple.client.server_system_time
+          uptime = simple.client.server_system_time - simple.client.server_start_time
+          info[:uptime] = seconds_to_timespan(uptime)
+        end
+        info[:server_guid] = simple.client.server_guid
+      end
 
-        #
-        # Create a descriptive string for service.info
-        #
-        desc = res['os'].dup
+      info[:preferred_dialect] = preferred_dialect
+      info[:versions] << version unless protocol.nil?
+    end
 
-        if res['edition'].to_s.length > 0
-          desc << " #{res['edition']}"
-          conf[:os_edition] = res['edition']
-          match_conf['os.edition'] = res['edition']
-        end
+    info
+  end
 
-        if res['sp'].to_s.length > 0
-          desc << " #{res['sp'].downcase.gsub('service pack ', 'SP')}"
-          conf[:os_sp] = res['sp']
-          match_conf['os.version'] = res['sp']
-        end
+  def smb_os_description(res, nd_smb_fingerprint)
+    #
+    # Create the note hash for fingerprint.match
+    #
+    nd_fingerprint_match = {}
+
+    #
+    # Create a descriptive string for service.info
+    #
+    desc = res['os'].dup
+
+    if !res['edition'].to_s.empty?
+      desc << " #{res['edition']}"
+      nd_smb_fingerprint[:os_edition] = res['edition']
+      nd_fingerprint_match['os.edition'] = res['edition']
+    end
 
-        if res['build'].to_s.length > 0
-          desc << " (build:#{res['build']})"
-          conf[:os_build] = res['build']
-          match_conf['os.build'] = res['build']
-        end
+    if !res['sp'].to_s.empty?
+      desc << " #{res['sp'].downcase.gsub('service pack ', 'SP')}"
+      nd_smb_fingerprint[:os_sp] = res['sp']
+      nd_fingerprint_match['os.version'] = res['sp']
+    end
+
+    if !res['build'].to_s.empty?
+      desc << " (build:#{res['build']})"
+      nd_smb_fingerprint[:os_build] = res['build']
+      nd_fingerprint_match['os.build'] = res['build']
+    end
+
+    if !res['lang'].to_s.empty? && res['lang'] != 'Unknown'
+      desc << " (language:#{res['lang']})"
+      nd_smb_fingerprint[:os_lang] = res['lang']
+      nd_fingerprint_match['os.language'] = nd_smb_fingerprint[:os_lang]
+    end
+
+    if simple.client.default_name
+      desc << " (name:#{simple.client.default_name})"
+      nd_smb_fingerprint[:SMBName] = simple.client.default_name
+      nd_fingerprint_match['host.name'] = nd_smb_fingerprint[:SMBName]
```

Comment:
Just for context, what's the semantic difference between `nd_smb_fingerprint` and `nd_fingerprint_match` ? Why does the key's naming convention differ? What does the `nd_` prefix mean?

Reply:
Those are the two different hashes that are passed to `report_note`. This code predates my changes but from what I can tell notes of `ntype: 'fingerprint.match'` get `nd_fingerprint_match` while notes of `ntype: 'smb.fingerprint'` get `nd_smb_fingerprint`. The variables might benefit from some clarification I supposed but I would hesitate to change the keys around in case they're used in some kind of workflow that I'm unaware of.